enhancements/proposals/node-autoreplacement.md

```diff
@@ -0,0 +1,128 @@
+# Node Autoreplacement
+
+## Summary
+
+Node Autoreplacement enhancement extends Scylla Operator with a functionality of automated replacement of nodes which lost their storage, which, at this point, renders them getting stuck on joining the cluster.
+
+This proposal suggests creating a module responsible for triggering a node replacement in such circumstances.
+As a side effect, introducing such functionality would allow for simplifying the existing `OrphanedPVController` by
+reducing its task to evicting orphaned PVs.
+
+## Motivation
+
+Currently, when a node looses its storage, it gets stuck on joining a cluster, since Scylla produces a following runtime error:
+```
+Startup failed: std::runtime_error (A node with address <ip> already exists, cancelling join. Use replace_address if you want to replace this node.)
+```
+
+Successfully restarting the node requires a manual intervention of a human operator by annotating the
+service corresponding to the affected node with a `scylla/replace` label. The main goal of this proposal is to improve
+Scylla Operator's stability in running Scylla independently of its administrators' interventions.
+
+### Goals
+
+- Successfully discovering and autoreplacing nodes which lost their storage without requiring user intervention.
+
+### Non-Goals
+
+- Removing orphaned PVs. This is left for `OrphanedPVController`.
+
+## Proposal
+
+Extend ScyllaCluster controller with code responsible for checking whether a given Scylla node:
+1. is **not** a first bootstrapping node in the cluster
+2. already exists in gossip
+
+In case both of these conditions are true, a dedicated ConfigMap is created/updated.
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: <service_name>-autoreplacement-config
+  ...
+data:
+  autoreplacement.config: |
+    replace-address-first-boot: <clusterIP>    
+```
+
+The config is later appended to the generic config during Scylla setup, which effectively triggers running Scylla with `replace-address-first-boot` flag.
+The ConfigMap takes precedence over the generic config. In case `replace-address-first-boot` is already defined there, it is overwritten.
+
+The ConfigMap is optional and unavailable during the first start of a pod. It is not live reloaded and only read during the entrypoint setup, after every restart of a pod.
+The ScyllaCluster controller talks to the nodes whenever a generated ConfigMap differs from the existing one, which essentially comes down to two cases: after first boot and after a change to corresponding service's `ClusterIP`.
```

Comment:
Won't the ConfigMaps differ through out the whole bootstrap process? That's currently like 1-2 minutes per node. Making manager API calls on every sync during that time is gonna be extremely costly. We need some condition that's based on the API object - e.g. could we only create the CM for nodes that are ready (and update always)? Guess we might not need to investigate the gossip at all then because a node ready implies presence in gossip.